### PR TITLE
**Feature:** Improve tree

### PR DESCRIPTION
**Pull request by fabien0102**
# Summary

- Indent more if the parent have an Icon
- Cut long items

# Related issue

# To be tested

Me
- [ ] No error or warning in the console on `localhost:6060`

Tester 1

- [ ] Things look good on the demo.
  

Tester 2

- [ ] Things look good on the demo.

### CODE REVIEW

File: src/Tree/TreeItem.tsx [hunk 1]
```diff
@@ -37,7 +40,7 @@ const Header = styled.div<{
   background-color: ${({ highlight, theme }) => (highlight ? theme.color.highlight : "none")};
   margin: 0 -${({ theme }) => theme.space.element}px;
   padding: ${({ theme }) => `${theme.space.base}px ${theme.space.element}px`};
-  padding-left: ${({ theme, level }) => theme.space.element * (level + 1)}px;
+  padding-left: ${({ theme, level, hasIconOffset }) => theme.space.element * (level + 1) + (hasIconOffset ? 24 : 0)}px;
```
Comment (ImogenF) [inline src/Tree/TreeItem.tsx]: Where does 24 come from?

Comment (fabien0102) [reply src/Tree/TreeItem.tsx]: Size of the icon + padding (eg: tweaking in the dev tool to be aligned ^^)

File: src/Tree/TreeItem.tsx [hunk 2]
```diff
@@ -148,16 +176,34 @@ const TreeItem: React.SFC<TreeItemProps> = ({
...
+        <DotMenuHorizontalIcon
+          size={18}
+          left
+          onClick={() => {
+            /** Just the hover style! */
+          }}
+          onMouseEnter={() => {
+            if (labelRef.current) {
+              const { left, top } = labelRef.current.getBoundingClientRect()
+              setViewMorePopup({ y: top + 20, x: left, content: label })
```
Comment (ImogenF) [inline src/Tree/TreeItem.tsx]: Why `+ 20`?

File: src/Tree/TreeItem.tsx (hunk 1 above)
Comment (ImogenF) [reply src/Tree/TreeItem.tsx]: Is there any way we can generalise this?

File: src/Tree/TreeItem.tsx (hunk 2 above)
Comment (fabien0102) [reply src/Tree/TreeItem.tsx]: Design requirement, it's not hide the label ;)

**Comment by ImogenF** [comment]
I don't know if this was part of the design, but the location of the popup looks strange to me, because it's so far away from the 3 dots that you hover over. 

image

Also, small thing: if you click on the 3 dots, they stay focused.

File: src/Tree/TreeItem.tsx [hunk 3]
```diff
@@ -148,16 +176,34 @@ const TreeItem: React.SFC<TreeItemProps> = ({
...
           highlightStyle={{ color: constants.color.text.action, backgroundColor: "transparent", fontWeight: "bold" }}
           searchWords={searchWords}
         />
       </Label>
-      <ActionsContainer highlight={highlight} childrenCount={React.Children.count(actions)}>
-        {actions}
-      </ActionsContainer>
+      {isTooLong && (
+        <DotMenuHorizontalIcon
+          size={18}
```
Comment (stereobooster) [inline src/Tree/TreeItem.tsx]:
```suggestion
          size={18}
          tabIndex={-1}
```

**Comment by stereobooster** [comment]
Screenshot 2019-08-21 at 14 39 15

File: src/Tree/TreeItem.tsx [hunk 4]
```diff
@@ -62,15 +66,21 @@ const Header = styled.div<{
   }
 `
 
+const viewMoreIconSize = 18
+
 const Label = styled("div")<{ hasChildren: boolean }>`
```
Comment (stereobooster) [inline src/Tree/TreeItem.tsx]:
```suggestion
const Label = styled.div<{ hasChildren: boolean }>`
```

File: src/Tree/TreeItem.tsx [hunk 5]
```diff
@@ -62,15 +66,21 @@ const Header = styled.div<{
...
+  /* Split the label by caract properly and show the first line only */
   overflow-wrap: break-word;
+  overflow: hidden !important;
```
Comment (stereobooster) [inline src/Tree/TreeItem.tsx]: why `!important`?

File: src/Tree/TreeItem.tsx [hunk 6]
```diff
@@ -121,16 +132,36 @@ const TreeItem: React.SFC<TreeItemProps> = ({
...
+  useLayoutEffect(() => {
+    // We can't attach the ref to `Highlighter`, this is why we attached the ref
+    // to the parent and using `children[0]`
+    if (labelRef.current && labelRef.current.children[0]) {
+      const { height } = labelRef.current.children[0].getBoundingClientRect()
+      if (height > 16) {
+        setIsTooLong(true)
+      }
+    }
+  })
```
Comment (stereobooster) [inline src/Tree/TreeItem.tsx]: do we want to run it on every re-render?

File: src/Tree/TreeItem.tsx [hunk 7]
```diff
@@ -121,16 +131,36 @@ const TreeItem: React.SFC<TreeItemProps> = ({
...
 
+  const [viewMorePopup, setViewMorePopup] = useState<{ x: number; y: number; content: string } | null>(null)
+  const [isTooLong, setIsTooLong] = useState(false)
+  const labelRef = useRef<HTMLDivElement>(null)
+  useLayoutEffect(() => {
+    // We can't attach the ref to `Highlighter`, this is why we attached the ref
+    // to the parent and using `children[0]`
+    if (labelRef.current && labelRef.current.children[0]) {
+      const { height } = labelRef.current.children[0].getBoundingClientRect()
+      if (height > 16) {
```
Comment (stereobooster) [inline src/Tree/TreeItem.tsx]:
```tsx
setIsTooLong(height > 16)
```

**Comment by stereobooster** [comment]
Dots are focusable but don't respond to keyboard actions: resolution either add tabIndex={-1} or handle keyboard.

**Pull request merged**